## 📚 내용 정리
### 자바에서 매우 중요한 예외
- 예외적인 일이 발생하면 `예외`라는 것을 던진다.


### try-catch는 짝이다.
```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 5 
    at c.exception.ExceptionSample.arrayOutOfBounds(ExceptionSample.java:11) 
    at c.exception.ExceptionSample.main(ExceptionSample.java:7)
```
- ArrayIndexOutOfBoundsException 예외 발생
- at으로 시작하는 스택 호출 추적(call stack trace, stack trace) 표시

- try-catch 에서 예외가 발생하지 않을 경우 try 내에 있는 모든 문장이 실행되고 try-catch 문장 이후 내용이 실행된다.
- try-catch 에서 예외가 발생하는 경우 try 내에서 예외가 발생한 이후의 문장들은 실행되지 않는다. catch 내에 있는 문장은 반드시 실행되고, try-catch 문장 이후의 내용이 실행된다.


### try-catch를 사용하면서 처음에 적응하기 힘든 변수 선언
- try블록 내에서 선언한 변수를 catch에서 사용할 수 없다. 일반적으로 catch 문장에서 사용할 변수는 try 앞에 미리 선언한다.

```java
public void checkVariable3() { 
	int[] intArray = null; 
	
	try { intArray = new int[5]; System.out.println(intArray[5]); 
	} catch (Exception e) { 
		System.out.println(intArray.length); 
	} 
	
	System.out.println("This code must run."); 
}
```


### finally야~ 넌 무슨 일이 생겨도 반드시 실행해야 돼
- finally 블록은 예외 발생 여부와 상관 없이 실행된다.


### 두 개 이상의 catch
- try 다음에 오는 catch 블록은 1개 이상 올 수 있다.
- 먼저 선언한 catch 블록의 예외 클래스가 다음에 선언한 catch 블록의 부모에 속하면, 자식에 속하는 catch 블록은 절대 실행이 안돼 컴파일 에러 발생
- 하나의 try 블록에서 예외가 발생하면 그 예외와 관련 있는 catch 블록을 찾아서 실행한다.
- catch 블록 중 발생한 예외와 관련있는 블록이 없으면, 예외가 발생하면서 쓰레드가 끝난다. 마지막 catch 블록에서 Exception 클래스로 묶어주면 방지 가능.


### 예외의 종류는 세 가지다.
- checked exception
- error
- runtime exception or unchecked exception

#### error
- 자바 프로그램 밖에서 발생한 예외
- Error는 프로세스에 영향을 주고, Exception은 쓰레드에만 영향을 준다.

#### runtime exception
- 런타임 예외는 예외가 발생할 것을 미리 감지하지 못했을 때 발생한다.
- RuntimeException을 확장한 예외들이다.
- 컴파일 시점에 체크를 하지 않아 `uncheckek exception` 이라고 부른다.

![[Pasted image 20250619092245.png]]
- Exception을 바로 확장한 클래스들이 Checked 예외
- RuntimeException을 확장한 클래스들이 런타임 예외


### 모든 예외의 할아버지는 java.lang.Throwable 클래스이다.
- Exception과 Error 클래스는 Throwable을 상속 받는다.
- Throwable 생성자
	- Throwable() 
	- Throwable(String message) 
	- Throwable(String message, Throwable cause) 
	- Throwable(Throwable cause)

#### Throwable 클래스에 선언되어 있고, Exception 클래스에서 Overriding한 메소드들
- getMessage(): 예외 메시지를 String 형태로 제공.
- toString: getMessage() 보다 자세하게 예외 메시지, 예외 클래스 이름 String 형태로 제공
- printStackTrace(): 첫 줄에는 예외 메시지 출력, 두 번째 줄부터는 예외가 발생하게 된 메소드들의 호출 관계(스택 트래에스)를 출력


### 난 예외를 던질 거니까 throws 라고 써 놓을께
- throw를 사용해 예가 발생된 메소드를 호출한 메소드로 던진다.
- throws를 명시한 메소드는 사용 시 try-catch 로 묶어줘야한다.

- 메소드 선언 시 매개 변수 소괄호 뒤에 throws 라는 예약어를 적어 준 뒤 예외를 선언하면, **해당 메소드에**서 선언한 예외가 발생했을 때 호출한 메소드로 예외가 전달된다. 두 가지 이상의 예외를 던질 수 있다면, 콤마로 구분해 예외 클래스 이름을 적어준다.
- try 블록 내에서 예외를 발생시킬 경우에는 throw 라는 예약어를 적어 준 뒤 예외 객체를 생성하거나, 생성되어있는 객체를 명시해준다. throw 한 예외 클래스가 catch 블록에 선언되어 있지 않거나, throws 선언에 포함되어 있지 않으면 컴파일 에러가 발생한다.
- catch 블록에서 예외를 throw 할 경우에도 메소드 선언의 throws 구문에 해당 예외가 정의되어 있어야한다.

```java
public void multiThrows() throws NullPointerException, Exception { }
```


### 내가 예외를 만들 수도 있다구?
- Error가 아닌 Exception을 처리하는 예외 클래스는 만들 수 있다.
- Throwable 이나 그 자식 클래스를 상속을 받는다.


### 자바 예외 처리 전략
![[Pasted image 20250619202035.png]]

- 예외 클래스를 만들 때에는, 반드시 try-catch로 묶어줄 필요가 있을 경우에만 Exception 클래스를 확장한다. 일반적으로 실행시 예외를 처리할 수 있는 경우에는 RuntimeException 클래스를 확장하는 것을 권장한다.



---
## ✨ 새로 알게된 내용


---
## ⛑️ 보충 필요
- Error는 프로세스에 영향을 주고, Exception은 쓰레드에만 영향을 준다. 각 차이점
- 구글에서 자바 예외 전략, Java Exception Strategy 확인해보기